lang/cs/Org.Apache.REEF.Common.Tests/Telemetry/MetricTests.cs

```diff
@@ -0,0 +1,106 @@
+using Org.Apache.REEF.Tang.Implementations.Tang;
+using Xunit;
+
+namespace Org.Apache.REEF.Common.Tests.Telemetry
+{
+    public class MetricsTests
+    {
+        /// <summary>
+        /// Test IMetrics, ICounters and IEvaluatorMetrics API.
+        /// </summary>
+        [Fact]
+        public void TestEvaluatorMetricsCountersOnly()
+        {
+            var evalMetrics1 = TangFactory.GetTang().NewInjector().GetInstance<IEvaluatorMetrics>();
+            var metrics1 = evalMetrics1.GetMetricsData();
+            metrics1.TryRegisterMetric(new Counter("counter1", "counter1 description"));
+            metrics1.TryRegisterMetric(new Counter("counter2", "counter2 description"));
+            ValidateMetric(metrics1, "counter1", 0);
+            ValidateMetric(metrics1, "counter2", 0);
+            for (int i = 0; i < 5; i++)
+            {
+                metrics1.Update("counter1", i);
+                metrics1.Update("counter2", i * 2);
+            }
+            ValidateMetric(metrics1, "counter1", 4);
+            ValidateMetric(metrics1, "counter2", 8);
+
+            var counterStr = metrics1.Serialize();
+
+            var evalMetrics2 = new EvaluatorMetrics(counterStr);
+            var metrics2 = evalMetrics2.GetMetricsData();
+            ValidateMetric(metrics2, "counter1", 4);
+            ValidateMetric(metrics2, "counter2", 8);
+        }
+
+        /// <summary>
+        /// Tests updating metric value.
+        /// </summary>
+        [Fact]
+        public void TestMetricSetValue()
+        {
+            var metrics = CreateMetrics();
+            metrics.TryRegisterMetric(new IntegerGauge("int1", "metric of type int", DateTime.Now.Ticks, 0));
+            metrics.TryRegisterMetric(new DoubleGauge("dou2", "metric of type double", DateTime.Now.Ticks, 0));
+            ValidateMetric(metrics, "int1", default(int));
+            ValidateMetric(metrics, "dou2", default(double));
+
+            metrics.Update(new IntegerGauge("int1", "new description", DateTime.Now.Ticks, 3));
+            metrics.Update("dou2", 3.14);
+
+            ValidateMetric(metrics, "int1", 3);
+            ValidateMetric(metrics, "dou2", 3.14);
+        }
+
+        /// <summary>
+        /// Test TryRegisterCounter with a duplicated counter name
+        /// </summary>
+        [Fact]
+        public void TestDuplicatedNames()
+        {
+            var metrics = CreateMetrics();
+            metrics.TryRegisterMetric(new Counter("metric1", "metric description"));
+            Assert.False(metrics.TryRegisterMetric(new Counter("metric1", "duplicate name")));
+        }
+
+        [Fact]
+        public void TestUpdateMetricWithDifferentType()
+        {
+            var metrics = CreateMetrics();
+            metrics.TryRegisterMetric(new IntegerGauge("int1", "metric of type int", DateTime.Now.Ticks, 0));
+            Assert.Throws<ApplicationException>(() => metrics.Update(new DoubleGauge("int1", "new description", DateTime.Now.Ticks, 3)));
```

Comment:
Do we want to add a new object every time an update occurs? I think the code should register a metric, get an instance to that metric and then set the value of that metric. Allocating a new object every time can be a large cost -- especially if doing so in a loop, etc.

Reply:
In the case where we want to keep a history of the metric, a new object would have to be created at each update anyway, so this would save cost only in the case where we are not interested in keeping a record.. adding this comment to narrow the scope.

Reply:
@singlis has a point in regards to recreating the object to pass around the "name" and "description" which is redundant. If we could just pass around the next "T value" that would be ideal.

Reply:
Changed so MetricData tracks updates directly made on the metric object. MetricData will create a MetricRecord which holds the value and timestamp to track updates, instead of copying the whole metric each time.